iree/hal/metal/metal_command_buffer.mm

```diff
@@ -81,12 +172,36 @@
                                       device_size_t length, const void* pattern,
                                       size_t pattern_length) {
   IREE_TRACE_SCOPE0("MetalCommandBuffer::FillBuffer");
-  return UnimplementedErrorBuilder(IREE_LOC) << "MetalCommandBuffer::FillBuffer";
+  IREE_ASSIGN_OR_RETURN(auto* target_device_buffer, CastBuffer(target_buffer));
+
+  target_offset += target_buffer->byte_offset();
+
+#ifdef IREE_PLATFORM_MACOS
+  // Per the spec for fillBuffer:range:value: "The alignment and length of the range must both be a
+  // multiple of 4 bytes in macOS, and 1 byte in iOS and tvOS."
+  if (target_offset % 4 != 0) {
+    return UnimplementedErrorBuilder(IREE_LOC)
+           << "MetalCommandBuffer::FillBuffer with offset that is not a multiple of 4 bytes";
+  }
+#endif
+
+  // Note that fillBuffer:range:value: only accepts a single byte as the pattern but FillBuffer
+  // can accept 1/2/4 bytes. If the pattern itself contains repeated bytes, we can call into
+  // fillBuffer:rnage:value:. Otherwise we may need to find another way. Leave that as
+  // unimplemented for now.
+  IREE_ASSIGN_OR_RETURN(uint8_t byte_pattern, GetRepeatedBytePattern(pattern, pattern_length));
```

Comment:
this is another big landmine that we should definitely avoid. Filling a buffer with floats is total valid and common, and I don't want some poor soul to come across this line when they first try to fill a buffer with anything but 0.0f. Special casing repeated values or zero as an optimization is fine, but not supporting some values is too scary. We can work around this limitation with a custom compute dispatch that does the fill when needed, and if you want to defer that I'd rather *only* have the case we will always be able to support here (when the pattern is 1 byte or 0) instead (so no checking for repeated values/etc - `if 0 or size==1 {fill} else {unimplemented}` - that'll make it much easier to spot and implement correctly when hit (vs happening to work until floating point rounding ends up with an initial value being 0.000001f :)

Reply:
Haha, sure! Removed the optimization path to only allow size-1 pattern atm.